File: leo2-ui/src/main/java/org/deku/leo2/bridge/LeoBridge.java
package org.deku.leo2.bridge;

import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider;
import org.deku.leo2.bridge.services.MessageService;
import org.glassfish.grizzly.http.server.HttpServer;
import org.glassfish.jersey.client.ClientProperties;
import org.glassfish.jersey.client.proxy.WebResourceFactory;
import org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory;
import org.glassfish.jersey.jackson.JacksonFeature;
import org.glassfish.jersey.server.ResourceConfig;
import sx.Disposable;
import sx.util.EventDelegate;
import sx.util.EventDispatcher;
import sx.util.EventListener;
import sx.util.ThreadSafeEventDispatcher;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import java.io.IOException;
import java.net.URI;

/**
 * Created by masc on 26.09.14.
 */
public class LeoBridge implements Disposable, MessageService.Listener {
    private static LeoBridge mInstance;

    public interface Listener extends EventListener {
        void onLeoBridgeMessageReceived(Message message);
    }

    /**
     * Created by masc on 23.07.14.
     */
    private class WebserviceResourceConfig extends ResourceConfig {
        public WebserviceResourceConfig()
        {
            super(JacksonFeature.class);

            // Server debug logging
            // registerInstances(new LoggingFilter(Logger.getLogger(LeoBridge.class.getName()), true));

            packages("org.deku.leo2.bridge.services");
        }
    }

    private static final URI HOST_URI = URI.create("http://localhost:37420/");
    private static final URI CLIENT_URI = URI.create("http://localhost:37421/");
    private HttpServer mHttpServer;
    private IMessageService mMessageServiceClient;

    EventDispatcher<Listener> mListenerEventDispatcher = new ThreadSafeEventDispatcher();
    public EventDelegate<Listener> getListenerEventDelegate() {
        return mListenerEventDispatcher;
    }

    public static LeoBridge instance() {
        if (mInstance == null) {
            synchronized (LeoBridge.class) {
                mInstance = new LeoBridge();
            }
        }
        return mInstance;
    }

    private LeoBridge() {

    }

    /**
     * Start leo bridge
     * @throws IOException
     */
    public void start() throws IOException {
        if (mHttpServer == null) {
            mHttpServer = GrizzlyHttpServerFactory.createHttpServer(HOST_URI, new WebserviceResourceConfig());
            // Setup mClient
            Client c = ClientBuilder.newClient();
            c.register(JacksonJsonProvider.class);
            c.property(ClientProperties.CONNECT_TIMEOUT, 500);

            // Client debug logging
            // c.register(new LoggingFilter(Logger.getLogger(LeoBridge.class.getName()), true));

            mMessageServiceClient = WebResourceFactory.newResource(IMessageService.class, c.target(CLIENT_URI));
        }
        mHttpServer.start();
    }

    /**
     * Stop leo bridge
     */
    public void stop() {
        if (mHttpServer != null)
            mHttpServer.shutdownNow();
    }

    public void sendMessage(Message message) {
        mMessageServiceClient.send(message);
    }

    public void sendValue(Object value) {
        mMessageServiceClient.send(new Message(value));
    }

    @Override
    public void onLeoBridgeServiceMessageReceived(Message message) {
        mListenerEventDispatcher.emit(r -> r.onLeoBridgeMessageReceived(message));
    }

    @Override
    public void dispose() {
        this.stop();
    }
}

File: leoz-ui/src/main/java/org/deku/leoz/bridge/LeoBridge.kt
package org.deku.leoz.bridge

import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider
import org.deku.leoz.bridge.services.MessageService
import org.glassfish.grizzly.http.server.HttpServer
import org.glassfish.jersey.client.ClientProperties
import org.glassfish.jersey.client.proxy.WebResourceFactory
import org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory
import org.glassfish.jersey.jackson.JacksonFeature
import org.glassfish.jersey.server.ResourceConfig
import sx.Disposable
import sx.event.EventDelegate
import sx.event.EventDispatcher
import sx.event.EventListener
import sx.event.ThreadSafeEventDispatcher
import java.io.IOException
import java.net.URI
import javax.ws.rs.client.ClientBuilder

/**
 * Created by masc on 26.09.14.
 */
class LeoBridge private constructor() : Disposable, MessageService.Listener {

    interface Listener : EventListener {
        fun onLeoBridgeMessageReceived(message: Message)
    }

    /**
     * Created by masc on 23.07.14.
     */
    private inner class WebserviceResourceConfig : ResourceConfig(JacksonFeature::class.java) {
        init {

            // Server debug logging
            // registerInstances(new LoggingFilter(Logger.getLog(LeoBridge.class.getName()), true));

            packages("org.deku.leoz.bridge.services")
        }
    }

    private var httpServer: HttpServer? = null
    private var messageServiceClient: IMessageService? = null

    internal var listenerEventDispatcher: EventDispatcher<Listener> = ThreadSafeEventDispatcher()
    val listenerEventDelegate: EventDelegate<Listener>
        get() = listenerEventDispatcher

    /**
     * Start leo bridge
     * @throws IOException
     */
    @Synchronized @Throws(IOException::class)
    fun start() {
        if (httpServer == null) {
            httpServer = GrizzlyHttpServerFactory.createHttpServer(HOST_URI, WebserviceResourceConfig())
            // Setup mClient
            val c = ClientBuilder.newClient()
            c.register(JacksonJsonProvider::class.java)
            c.property(ClientProperties.CONNECT_TIMEOUT, 500)

            // Client debug logging
            // c.register(new LoggingFilter(Logger.getLog(LeoBridge.class.getName()), true));

            messageServiceClient = WebResourceFactory.newResource(IMessageService::class.java, c.target(CLIENT_URI))
        }
        httpServer!!.start()
    }

    /**
     * Stop leo bridge
     */
    @Synchronized fun stop() {
        if (httpServer != null)
            httpServer!!.shutdownNow()
    }

    fun sendMessage(message: Message) {
        messageServiceClient!!.send(message)
    }

    fun sendValue(value: Any) {
        messageServiceClient!!.send(Message(value))
    }

    override fun onLeoBridgeServiceMessageReceived(message: Message) {
        listenerEventDispatcher.emit { r -> r.onLeoBridgeMessageReceived(message) }
    }

    override fun close() {
        this.stop()
    }

    companion object {
        private var instance: LeoBridge? = null

        private val HOST_URI = URI.create("http://localhost:37420/")
        private val CLIENT_URI = URI.create("http://localhost:37421/")

        fun instance(): LeoBridge {
            if (instance == null) {
                synchronized (LeoBridge::class.java) {
                    instance = LeoBridge()
                }
            }
            return instance!!
        }
    }
}
